### ISSUE TITLE
Error 2 in installation process with no other hints as to what has gone wrong

### ISSUE THREAD

**Issue by jprattgsu** [trilinos/Trilinos#4277]
@trilinos/<teamName>

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by jprattgsu** [comment]
I have a new Intel Skylake cluster with Intel 18 compilers.  This cluster is now 2 months old, and for 2 months I have been struggling with installing Trilinos.  I'd be grateful for any ideas for how to get further because at this point I am not making forward progress.  I'm working with the latest version from github, although I've also tried versions 11.4.3, 12.10.1, and 12.12.1.  I am getting an Error 2 late in the build process, an error that is not easily searchable.  Here is the final output and error statement:

[ 76%] Built target ThyraCore_DiagonalQuadraticResponseOnlyModelEvaluatorUnitTests
[ 76%] Linking CXX executable ThyraCore_Simple2DModelEvaluatorUnitTests.exe
cd /home/jpratt/bin/trilinosbuildgitrepo/packages/thyra/core/test/nonlinear/models/UnitTests && /usr/bin/cmake3 -E cmake_link_script CMakeFiles/ThyraCore_Simple2DModelEvaluatorUnitTests.dir/link.txt --verbose=1
/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/mpiicpc   -std=c++11 -O3 -DNDEBUG   CMakeFiles/ThyraCore_Simple2DModelEvaluatorUnitTests.dir/Simple2DModelEvaluator_UnitTests.cpp.o CMakeFiles/ThyraCore_Simple2DModelEvaluatorUnitTests.dir/__/__/__/__/__/__/teuchos/core/test/UnitTest/Teuchos_StandardUnitTestMain.cpp.o  -o ThyraCore_Simple2DModelEvaluatorUnitTests.exe ../libthyra_nonlinear_test_models.a ../../../../src/libthyracore.a ../../../../../../rtop/src/librtop.a ../../../../../../teuchos/numerics/src/libteuchosnumerics.a -L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/scalapack-2.0.2/ -lscalapack -lmpi -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -lz -lrt -lm -lpthread -L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -limf -lsvml -lintlc  -lifcore -lmpi -lz -lrt -lm -lpthread ../../../../../../teuchos/comm/src/libteuchoscomm.a ../../../../../../teuchos/parameterlist/src/libteuchosparameterlist.a ../../../../../../teuchos/parser/src/libteuchosparser.a ../../../../../../teuchos/core/src/libteuchoscore.a ../../../../../../kokkos/core/src/libkokkoscore.a /usr/lib64/libdl.so
make[2]: Leaving directory `/home/jpratt/bin/trilinosbuildgitrepo'
[ 76%] Built target ThyraCore_Simple2DModelEvaluatorUnitTests
[ 76%] Linking CXX executable ThyraEpetraAdapters_test_epetra_adapters.exe
cd /home/jpratt/bin/trilinosbuildgitrepo/packages/thyra/adapters/epetra/test && /usr/bin/cmake3 -E cmake_link_script CMakeFiles/ThyraEpetraAdapters_test_epetra_adapters.dir/link.txt --verbose=1
/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/mpiicpc   -std=c++11 -O3 -DNDEBUG   CMakeFiles/ThyraEpetraAdapters_test_epetra_adapters.dir/test_epetra_adapters.cpp.o  -o ThyraEpetraAdapters_test_epetra_adapters.exe ../src/libthyraepetra.a ../../../../triutils/src/libtriutils.a ../../../core/src/libthyracore.a ../../../../rtop/src/librtop.a ../../../../epetra/src/libepetra.a ../../../../teuchos/numerics/src/libteuchosnumerics.a ../../../../teuchos/kokkoscomm/src/libteuchoskokkoscomm.a ../../../../teuchos/comm/src/libteuchoscomm.a ../../../../teuchos/kokkoscompat/src/libteuchoskokkoscompat.a ../../../../teuchos/parameterlist/src/libteuchosparameterlist.a ../../../../teuchos/remainder/src/libteuchosremainder.a ../../../../teuchos/parser/src/libteuchosparser.a -L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/scalapack-2.0.2/ -lscalapack -lmpi -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -lz -lrt -lm -lpthread -L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -limf -lsvml -lintlc  -lifcore -lmpi -lz -lrt -lm -lpthread ../../../../teuchos/core/src/libteuchoscore.a ../../../../kokkos/core/src/libkokkoscore.a /usr/lib64/libdl.so -lifport -lifcoremt
make[2]: Leaving directory `/home/jpratt/bin/trilinosbuildgitrepo'
[ 76%] Built target ThyraEpetraAdapters_test_epetra_adapters
make[1]: Leaving directory `/home/jpratt/bin/trilinosbuildgitrepo'
make: *** [all] Error 2

When I make a second time, the make process passes through this particular problem area, but again ends in an Error 2 somewhat later.  (Configuration finished with no errors).

Here is what I am using in the do-configure file:

cmake3 \
-D TPL_ENABLE_BLAS=ON \
-D TPL_BLAS_LIBRARIES:STRING="-L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -limf -lsvml -lintlc  -lifcore -lmpi -lz -lrt -lm -lpthread" \
\
-D TPL_ENABLE_LAPACK=ON \
-D TPL_LAPACK_LIBRARIES:STRING="-L/opt/lapack-3.5.0/ -llapack -lrefblas -ltmglib -L/opt/scalapack-2.0.2/ -lscalapack -lmpi -L/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/ -lz -lrt -lm -lpthread" \
\
-D TPL_ENABLE_METIS:BOOL=ON \
-D METIS_LIBRARY_DIRS:PATH="/opt/metis-5.1.0/lib/" \
-D METIS_INCLUDE_DIRS:PATH="/opt/metis-5.1.0/include/" \
-D METIS_LIBRARY_NAMES:STRING="metis;z;rt;m;pthread" \
\
-D TPL_ENABLE_ParMETIS:BOOL=ON \
-D ParMETIS_INCLUDE_DIRS:PATH="/opt/parmetis-4.0.3/include" \
-D ParMETIS_LIBRARY_DIRS:PATH="/opt/parmetis-4.0.3/lib;/opt/metis-5.1.0/lib/" \
-D ParMETIS_LIBRARY_NAMES:STRING="parmetis;metis;z;rt;m;pthread" \
\
-D TPL_ENABLE_MUMPS:BOOL=ON \
-D MUMPS_INCLUDE_DIRS:PATH="/usr/include/MUMPS/" \
-D MUMPS_LIBRARY_DIRS:PATH="/opt/lapack-3.5.0/lib/;/opt/scalapack-2.0.2/lib/;/opt/metis-5.1.0/lib/;/opt/parmetis-4.0.3/lib/;/opt/metis-5.1.0/lib/" \
-D MUMPS_LIBRARY_NAMES:STRING="dmumps;mumps_common;pord;parmetis;metis;scalapack;tmglib;lapack;blas" \
\
-D TPL_ENABLE_MPI:BOOL=ON \
-D MPI_BASE_DIR:PATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/;/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/lib/intel64_lin/" \
-D MPI_LIBRARY_DIR:PATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/lib64/" \
-D MPI_BIN_DIR:PATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/" \
-D MPI_INCLUDE_PATH:PATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/include64/;/opt/intel/compilers_and_libraries_2018.3.222/linux/compiler/include/" \
-D MPI_EXEC:FILEPATH="mpirun" \
-D MPI_EXEC_MAX_NUMPROCS:STRING=4 \
\
-D CMAKE_INSTALL_PREFIX:PATH=${TRILINOS_INSTALL_PATH} \
-D CMAKE_BUILD_TYPE:STRING=RELEASE \
-D CMAKE_CXX_COMPILER:FILEPATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/mpiicpc" \
-D CMAKE_C_COMPILER:FILEPATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/mpiicc" \
-D CMAKE_Fortran_COMPILER:FILEPATH="/opt/intel/compilers_and_libraries_2018.3.222/linux/mpi/bin64/mpiifort" \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
-D HAVE_GCC_ABI_DEMANGLE:BOOL=ON \
-D BUILD_SHARED_LIBS:BOOL=${SHARED_LIB} \
\
-D Trilinos_WARNINGS_AS_ERRORS_FLAGS:STRING="" \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=ON \
-D Trilinos_ENABLE_ALL_PACKAGES:BOOL=ON \
-D Trilinos_ENABLE_CTrilinos:BOOL=ON \
-D Trilinos_ENABLE_ForTrilinos:BOOL=ON \
-D Trilinos_ENABLE_SEACAS:BOOL=OFF \
-D Trilinos_ENABLE_Zoltan2:BOOL=OFF \
-D Trilinos_ENABLE_STK:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=ON \
-D Trilinos_ENABLE_EXAMPLES:BOOL=ON \
-D CTrilinos_ENABLE_TESTS:BOOL=OFF \
-D CTrilinos_ENABLE_EXAMPLES:BOOL=OFF \
-D ForTrilinos_ENABLE_TESTS:BOOL=ON \
-D ForTrilinos_ENABLE_EXAMPLES:BOOL=ON \
-D CMAKE_CXX_COMPILER_FLAGS:STRING="" \
-D Trilinos_EXTRA_LINK_FLAGS:STRING="" \
-DTPL_ENABLE_Netcdf=OFF \
-DTPL_ENABLE_Boost=OFF \
-DTPL_ENABLE_GLM=OFF \
-Wno-dev \
${EXTRA_ARGS} \
${TRILINOS_SRC_PATH}

**Comment by mhoemmen** [comment]
@trilinos/framework

**Comment by ndellingwood** [comment]
@jprattgsu I'm not sure if this will help, but a couple comments regarding changes of the configuration script to try out:

1. The BLAS_LIBRARIES string above looks like it's expecting lapack libraries, maybe copy/paste error?
`-D TPL_BLAS_LIBRARIES:STRING="-L/opt/lapack-3.5.0/ -llapack -lrefblas ...`

2. I'm not positive of CMake's requirements for path strings, but it may be worth removing any trailing "/" at the end of paths, 

e.g. the first instance I see with a trailing slash "/", 
`-D METIS_LIBRARY_DIRS:PATH="/opt/metis-5.1.0/lib/" `

try removing the trailing "/", that is 

`-D METIS_LIBRARY_DIRS:PATH="/opt/metis-5.1.0/lib" `

and doing this throughout.

**Comment by mhoemmen** [comment]
For BLAS and LAPACK settings, see the following document, linked in the Trilinos README when you view github.com/trilinos/Trilinos:

https://trilinos.org/docs/files/TrilinosBuildReference.html

```
WARNING: Do not try to hack the system and set, for example:

TPL_BLAS_LIBRARIES="-L/some/dir -llib1 -llib2 ..."
This is not compatible with proper CMake usage and it not guaranteed to be supported for all use cases or all platforms!
```
The proper way to do this is to set `BLAS_LIBRARY_DIRS` and `BLAS_LIBRARY_NAMES`.  Ditto for LAPACK.

**Comment by mhoemmen** [comment]
The point is that you should never set `TPL_BLAS_LIBRARIES`.  Do the following things, in order:

1. If you _only_ set `TPL_ENABLE_BLAS:BOOL=ON`, Trilinos will search system directories for the BLAS.
2. If you set that _and also_ set `BLAS_LIBRARY_DIRS`, Trilinos will search those directories for the BLAS, using standard library names (e.g., `-lblas`).
3. If you set (1) and (2), and also set `BLAS_LIBRARY_NAMES`, Trilinos will search the aforementioned directories, for the names that you specify.  Names should be without prefixes or suffixes.  For example: `BLAS_LIBRARY_NAMES="fancyblas;funnyblas;weirdblas"`.

**Comment by william76** [comment]
@jprattgsu I'm closing this issue b/c it's been inactive over a year. If you're still having this problem with the latest Trilinos please re-open or file a new issue. 😃 

FYI @ndellingwood (since you commented on this back in the day :p)
